sdk/core/azure-core/src/main/java/com/azure/core/util/Option.java

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
+/**
+ * The Option type, every Option is either has a value present including null,
+ * or has no value present.
+ *
+ * @param <T> The value type.
+ */
+public final class Option<T> {
+    private static final Option<?> UNSET = new Option<>();
+    private final boolean isSet;
+    private final T value;
+
+    /**
+     * Returns an {@link Option} instance with no value.
+     *
+     * @param <T> The value type.
+     * @return An Option type with no value present.
+     */
+    public static <T> Option<T> unset() {
```

Comment:
I think we should change this into `empty` and make the current `empty` `clear`. This pattern will follow more closely with what `Optional` types in Java do, where it is expected that `isPresent` on `empty` will return false. With the current design `empty` will return true for `isSet`, which makes it a better candidate to be called `clear`.

Reply:
The `Option` type is for "tri-state", to support `explicit-null`. When the user uses `Option::ofNullable(null)`, it means value-present as `explicit-null`.

The JDK `Optional` is for "two-sate". The `Option::ofNullable(non-null)` means value-present and `Option::ofNullable(null)` means value-not-present.  `Optional::ofNullable(null)`  == `Optional::empty()` (`isPresent() == false`).

Due to above "tri-state" vs "two-state", the meaning of `Option::ofNullable(null)` [presense of explicit-null] and `Optional::ofNullable(null)` [absense] is different.

Now I wonder should we remove `Option::empty()` from the"Option" to reduce confusion? then we'll have:

  `Option.of(non-null)` | `Option.ofNullable(value|null)` == value-present
  `Option.unset()` == value-not-present


Btw I'm still not happy with the name `Option.unset()`, while `Option` a normal type but when used in the REST API context, `Option.unset()` means don't change the server value.

Thoughts? 

\\cc @alzimmermsft  @JonathanGiles @srnagar

Reply:
We can borrow the JavaScript terminology for tri-state - `uninitialized` for to represent not set, `empty` for `null` value and `T` for non-null value.

Reply:
I think that's what we have today, js_uninitialized == java_unset, Option.empty() == explicit-null. One proposed name for "unset" is "uninitialized" i.e. `Option.uninitialized()`.

Regarding using "empty" for "null", do want `empty`? or given Option is for "tri-state" and "null" is a valid value, should we just expose "Option.ofNullable(null)" for explicit-null?

Reply:
Yeah, I was proposing we use the term `uninitialized` instead of `unset`. `unset` can have dual meaning which can either mean that the value is not set or that the value that is already set is reverted.

Reply:
renamed `unset` to `uninitialized`.

I kept `empty()` with the same meaning i.e. `empty` == initialized with a null